activerecord/lib/active_record/relation.rb

```diff
@@ -67,6 +67,11 @@ def bind_attribute(name, value) # :nodoc:
     #   user = users.new { |user| user.name = 'Oscar' }
     #   user.name # => Oscar
     def new(attributes = nil, &block)
+      current_scope = klass.current_scope(true)
```

Comment:
It feels to me that this is being done at the wrong level. Relation should not know about scoping. Maybe this should be done in the scoping class?

cc @matthewd

Reply:
I agree that here is not best place to do this.
But Relation has `scoping` method, and already need to know the result of `klass.current_scope(true)` in some places (in `scoping`, `spawn`, etc), so I just do this here since here is only place to do this.

If we'd not like to do this here, an option is just extracting to method in the scoping class like the following:

```diff
diff --git a/activerecord/lib/active_record/relation.rb b/activerecord/lib/active_record/relation.rb
index 32f0609798..347d745d19 100644
--- a/activerecord/lib/active_record/relation.rb
+++ b/activerecord/lib/active_record/relation.rb
@@ -67,11 +67,7 @@ def bind_attribute(name, value) # :nodoc:
     #   user = users.new { |user| user.name = 'Oscar' }
     #   user.name # => Oscar
     def new(attributes = nil, &block)
-      current_scope = klass.current_scope(true)
-      block = -> record do
-        klass.current_scope = current_scope
-        yield record if block_given?
-      end
+      block = klass.current_scope_restoring_block(&block)
       scoping { klass.new(attributes, &block) }
     end
 
diff --git a/activerecord/lib/active_record/scoping.rb b/activerecord/lib/active_record/scoping.rb
index 35e9dcbffc..1142a87d25 100644
--- a/activerecord/lib/active_record/scoping.rb
+++ b/activerecord/lib/active_record/scoping.rb
@@ -30,6 +30,14 @@ def current_scope(skip_inherited_scope = false)
       def current_scope=(scope)
         ScopeRegistry.set_value_for(:current_scope, self, scope)
       end
+
+      def current_scope_restoring_block(&block)
+        current_scope = self.current_scope(true)
+        -> *args do
+          self.current_scope = current_scope
+          yield(*args) if block_given?
+        end
+      end
     end
 
     def populate_with_current_scope_attributes # :nodoc:
```

Reply:
That could be a nice refactoring

Reply:
Applied 4cb1438b57067d637c79d49d0662c43b5b4e64c2.